### ISSUE TITLE
signature for callOutViewFor:annotation should have MGLCalloutView return type

### ISSUE THREAD

**Issue by idrougge** [mapbox/mapbox-gl-native#8420]
**Platform:** Swift
**Mapbox SDK version:** 3.4.1

`func mapView(_ mapView: MGLMapView, calloutViewFor annotation: MGLAnnotation) -> UIView?` as a signature is misleading since the function expects the return `UIView` to conform to `MGLCalloutView` and will cause a crash because of sending incorrect selectors. The return type should be changed to `MGLCalloutView` in order to reflect this.

**Comment by 1ec5** [comment]
Retitling, because macOS is unaffected: `-[MGLMapViewDelegate mapView:calloutViewControllerForAnnotation:]` returns an NSViewController, and the return value is always treated as a plain NSViewController.

**Comment by boundsj** [comment]
This was fixed by https://github.com/mapbox/mapbox-gl-native/pull/8541 that landed in the `release-ios-v3.5.0-android-v5.0.0` branch and will be in the [ios v3.5.1 patch release](https://github.com/mapbox/mapbox-gl-native/milestone/57)

**Comment by boundsj** [comment]
Reopening.

Unfortunately, I missed that this change causes a warning in `MGLMapView`:

```Objective-C
calloutView = [self.delegate mapView:self calloutViewForAnnotation:annotation];
```

triggers:

```
/mapbox-gl-native/platform/ios/src/MGLMapView.mm:3764:27: Incompatible pointer types assigning to 'UIView<MGLCalloutView> *' from 'id<MGLCalloutView> _Nullable'
```

The naming of `MGLCalloutView` protocol and its documentation make the correct choice for what base type to return fairly obvious. However, I think we should guard against using objects that are not a kind of `UIView` and, if they are views, cast them as such to silence the warning.

cc @1ec5
